### PR TITLE
Eliminate calls with FILE_FLAG_OVERLAPPED set, but no overlapped

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Resolves: #3325

## Description

IOCTL calls from multiple threads sharing the same device handle that has FILE_FLAG_OVERLAPPED but without an OVERLAPPED can cause IOCTL calls to hang.

## Testing

CI/CD

## Documentation

No.

## Installation

No.

### CODE REVIEW

File: libs/api_common/device_helper.cpp [hunk 1]
```diff
@@ -26,7 +26,10 @@ typedef struct _async_ioctl_completion_context
     async_ioctl_completion_callback_t callback;
 } async_ioctl_completion_context_t;
 
+// Handle used for synchronous calls to the driver.
 static ebpf_handle_t _device_handle = ebpf_handle_invalid;
```
Comment (dthaler) [inline libs/api_common/device_helper.cpp:30]:
```suggestion
static ebpf_handle_t _sync_device_handle = ebpf_handle_invalid;
```

File: libs/api_common/device_helper.cpp [hunk 2]
```diff
@@ -71,21 +71,21 @@ clean_up_device_handle()
 {
```
Comment (dthaler) [inline libs/api_common/device_helper.cpp:71]: initialize_device_handle() initializes both _sync_device_handle and _async_device_handle.
Shouldn't clean_up_device_handle() then clean up both?

File: libs/api_common/device_helper.cpp [hunk 3]
```diff
@@ -71,21 +71,21 @@ clean_up_device_handle()
...
-        Platform::CloseHandle(_device_handle);
-        _device_handle = ebpf_handle_invalid;
+    if (_sync_device_handle != ebpf_handle_invalid) {
+        Platform::CloseHandle(_sync_device_handle);
+        _sync_device_handle = ebpf_handle_invalid;
     }
 }
 
 ebpf_handle_t
 get_device_handle()
```
Comment (dthaler) [inline libs/api_common/device_helper.cpp:81]:
```suggestion
get_sync_device_handle()
```

File: libs/api_common/device_helper.hpp [hunk 4]
```diff
@@ -37,6 +37,9 @@ clean_up_device_handle();
 ebpf_handle_t
 get_device_handle();
```
Comment (dthaler) [inline libs/api_common/device_helper.hpp:38]:
```suggestion
get_sync_device_handle();
```

File: libs/api_common/device_helper.hpp [hunk 5]
```diff
@@ -101,7 +104,7 @@ invoke_ioctl(request_t& request, reply_t& reply = _empty_reply, _Inout_opt_ OVER
     }
 
     auto success = Platform::DeviceIoControl(
-        get_device_handle(),
+        overlapped ? get_async_device_handle() : get_device_handle(),
```
Comment (dthaler) [inline libs/api_common/device_helper.hpp:107]:
```suggestion
        overlapped ? get_async_device_handle() : get_sync_device_handle(),
```

**Comment by gtrevi** [comment]
Do we need integrate specific tests, similar to the ones that surfaced this issue?

**Review by matthewige — APPROVED** [review]

File: libs/api_common/device_helper.cpp (hunk 2 above)
Comment (dthaler) [reply libs/api_common/device_helper.cpp:71]: @Alan-Jowett Still waiting for an answer to this comment.

File: libs/api_common/device_helper.cpp [hunk 6]
```diff
@@ -26,22 +26,40 @@ typedef struct _async_ioctl_completion_context
     async_ioctl_completion_callback_t callback;
 } async_ioctl_completion_context_t;
 
-static ebpf_handle_t _device_handle = ebpf_handle_invalid;
+// Handle used for synchronous calls to the driver.
+static ebpf_handle_t _sync_device_handle = ebpf_handle_invalid;
```
Comment (saxena-anurag) [inline libs/api_common/device_helper.cpp:30]: do we need a handle per thread (TLS)? else all the sync IOCTL calls will be serialized.

File: libs/api_common/device_helper.cpp (hunk 2 above)
Comment (Alan-Jowett) [reply libs/api_common/device_helper.cpp:71]: Fixed in: https://github.com/microsoft/ebpf-for-windows/pull/3315/commits/9b0bac9741498df94fa117e9f33832da67daa65f

File: libs/api_common/device_helper.cpp (hunk 6 above)
Comment (Alan-Jowett) [reply libs/api_common/device_helper.cpp:30]: I tried that approach, but it runs into issues with our existing unit testing framework. Specifically, the TLS destroy occurs after the test completes and the closehandle mock has been remove. 

My proposed solution is to fix this the simple way for now, and then take the time to build a completer and more performant fix when we are not time bound.

**Review by dthaler — APPROVED** [review]

File: libs/api_common/device_helper.cpp (hunk 6 above)
Comment (shankarseal) [reply libs/api_common/device_helper.cpp:30]: I would recommend to modify `ebpf_clear_thread_local_storage` to keep the handle thing alive?

Comment (Alan-Jowett) [reply libs/api_common/device_helper.cpp:30]: Lets commit this as v1. We can then decide on a more performant v2.

**Pull request merged**